File: 프로그래머스/lv2/87946. 피로도/피로도.py
answer = -1
def solution(k, dungeons):
  
    n = len(dungeons)
    visited = [0] * n
    def backtrack(index, n, k):
        global answer
        if index > answer:
            answer = index
        for i in range(n):
            if k >= dungeons[i][0] and visited[i] == 0:
                visited[i] = 1
                k -= dungeons[i][1]
                backtrack(index + 1,  n, k)
                visited[i] = 0
                k += dungeons[i][1]
                
        
    backtrack(0, n, k)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    int answer;
    int n, k;
    int[][] dungeons;
    boolean[] visited;
    public int solution(int k, int[][] dungeons) {
        init(k, dungeons);
        solve();
        return answer;
    }
    
    void init(int k, int[][] dungeons) {
        this.answer = 0;
        this.n = dungeons.length;
        this.k = k;
        this.dungeons = dungeons;
        this.visited = new boolean[n];
    }
    
    void solve() {
        dfs(k, 0);
    }
    
    void dfs(int tired, int caught) {
        answer = Math.max(answer, caught);
        
        for (int i = 0; i < n; i++) {
            if (!visited[i] && tired >= dungeons[i][0]) {
                visited[i] = true;
                dfs(tired - dungeons[i][1], caught + 1);
                visited[i] = false;
            }
        }
    }
}